cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -236,35 +225,33 @@ private boolean checkJobConstraints(DatasetContext context) throws Exception {
           emptyScan = false;
           checkAndUpdateJob(jobQueue, job);
         }
-      } finally {
-        jobQueueIter.close();
       }
       return emptyScan;
     }
 
     private void checkAndUpdateJob(JobQueueDataset jobQueue, Job job) {
-      switch (job.getState()) {
-        case PENDING_TRIGGER:
-          if (isTriggerSatisfied(job)) {
-            jobQueue.transitState(job, Job.State.PENDING_CONSTRAINT);
-          }
-        // fallthru to check for constraint satisfaction after trigger satisfaction
-        case PENDING_CONSTRAINT:
-        case PENDING_LAUNCH:
-          if (constraintsSatisfied(job)) {
-            jobQueue.transitState(job, Job.State.PENDING_LAUNCH);
-            readyJobs.add(job);
-          }
-          break;
+      if (job.getState() == Job.State.PENDING_TRIGGER) {
+        if (!isTriggerSatisfied(job)) {
+          return;
+        }
+        job = jobQueue.transitState(job, Job.State.PENDING_CONSTRAINT);
+      }
+      // check for constraint satisfaction after trigger satisfaction
+      if (!constraintsSatisfied(job)) {
+        return;
+      }
+      if (job.getState() == Job.State.PENDING_CONSTRAINT) {
+        jobQueue.transitState(job, Job.State.PENDING_LAUNCH);
+      }
+      if (!readyJobs.contains(job)) {
+        readyJobs.add(job);
       }
     }
 
-    private boolean runReadyJobs(DatasetContext context) throws IOException, DatasetManagementException {
-      JobQueueDataset jobQueue = Schedulers.getJobQueue(context, datasetFramework);
-      Iterator<Job> readyJobsIter = readyJobs.iterator();
-      while (readyJobsIter.hasNext() && isRunning()) {
+    private boolean runReadyJobs() throws IOException, DatasetManagementException {
+      Iterator<Job> readyJobsIter = Iterators.consumingIterator(readyJobs.iterator());
+      while (readyJobsIter.hasNext() && !stopping) {
```

Comment:
this loop will lauch at most one job, then return true, correct? What if there are more than one that are ready?